### ISSUE TITLE
[3.0.0] OSSL_PARAM_utf8_string example

### ISSUE THREAD

**Issue by romen** [openssl/openssl#12506]
In `doc/man3/EVP_PKEY_fromdata.pod`:

https://github.com/openssl/openssl/blob/5ac582d949c4f0dbf919c99d59496035a1f7e982/doc/man3/EVP_PKEY_fromdata.pod#L151-L156

I see at least three issues in that example:
- `"curve-name"` became `"group"` (easy to fix)
- `OSSL_PARAM_utf8_string()`  expects 3 parameters
- this use of `OSSL_PARAM_utf8_string()`, even adding the length (or `0`), is problematic because the literal string is a `const char*`

Can someone help me figure out what this minimal example should look like?
Internally it seems we always use the OSSL_PARAM_BLD approach, so we should pay extra care in making sure that the documentation for external users interfacing with `OSSL_PARAM` APIs directly are error-free.

**Comment by romen** [comment]
Ping @mattcaswell , @paulidale , @slontis

**Comment by levitte** [comment]
Hmmm, #12389?

**Comment by levitte** [comment]
> - this use of `OSSL_PARAM_utf8_string()`, even adding the length (or `0`), is problematic because the literal string is a `const char*`

No sure I understand why that's a problem.  Am I missing something compiler related?

**Comment by levitte** [comment]
> Internally it seems we always use the OSSL_PARAM_BLD approach

Really?

**Comment by romen** [comment]
> > Internally it seems we always use the OSSL_PARAM_BLD approach
> 
> Really?

I was looking for examples in the tests where utf8_strings are used and I couldn't find any

**Comment by romen** [comment]
> Hmmm, #12389?

The updated example also uses the `OSSL_PARAM_BLD`!

**Comment by romen** [comment]
> > * this use of `OSSL_PARAM_utf8_string()`, even adding the length (or `0`), is problematic because the literal string is a `const char*`
> 
> No sure I understand why that's a problem. Am I missing something compiler related?

```
../../home/tuveri/repos/ossl_committer/test_ec_explicit_params_api/test/ectest.c:2630:13: error: initializing 'void *' with an expression of type 'const char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
            OSSL_PARAM_utf8_string("group", (const char*)curve_name, strlen(curve_name)),
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../home/tuveri/repos/ossl_committer/test_ec_explicit_params_api/include/openssl/params.h:60:5: note: expanded from macro 'OSSL_PARAM_utf8_string'                                                             
    OSSL_PARAM_DEFN((key), OSSL_PARAM_UTF8_STRING, (addr), sz)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../home/tuveri/repos/ossl_committer/test_ec_explicit_params_api/include/openssl/params.h:27:22: note: expanded from macro 'OSSL_PARAM_DEFN'                                                                   
    { (key), (type), (addr), (sz), OSSL_PARAM_UNMODIFIED }
                     ^~~~~~
1 error generated.
```

**Comment by paulidale** [comment]
There are quite a number of pitfalls to navigate trying to create statically allocated OSSL_PARAM arrays.

It is best to dynamically create them using `OSSL_PARAM_construct_*` or `OSSL_PARAM_BLD_*`.  Since you've got BIGNUM's, use `OSSL_PARAM_BLD_*`.  BIGNUMs are awkard otherwise.

**Comment by romen** [comment]
> There are quite a number of pitfalls to navigate trying to create statically allocated OSSL_PARAM arrays.
> 
> It is best to dynamically create them using `OSSL_PARAM_construct_*` or `OSSL_PARAM_BLD_*`. Since you've got BIGNUM's, use `OSSL_PARAM_BLD_*`. BIGNUMs are awkard otherwise.

So, is this the recommended API and we should discourage users from using `OSSL_PARAM`s directly?
Does our documentation reflect this?

**Comment by paulidale** [comment]
In my view there is a decreasing level of difficulty but an increasing level of over head for the three possibilities:

1. Direct use of `OSSL_PARAM`.
1. Using the `OSSL_PARAM_construct_xxx()` calls is good but BIGNUMs are rather tricky.  The overhead is slight.
1. Using the `OSSL_PARAM_BLD_xxx()` calls is easiest but has most overhead (extra memory allocations and copying).

If it were me, I'd used the `OSSL_PARAM_construct_xxx()` unless a BIGNUM were involved.  Then I'd use `OSSL_PARAM_BLD_xxx()`.  I would never voluntarily choose the raw `OSSL_PARAM` approach.

**Comment by romen** [comment]
So, given this answer and the fact that any external provider author has to deal with `OSSL_PARAM` strings even for the most basic setup operations, shouldn't the various `OSSL_PARAM_construct/BLD` functions be included in `core_dispatch`?

https://github.com/openssl/openssl/blob/79410c5f8b139c423be436810b4fe4de4637fc24/crypto/provider_core.c#L1043-L1088

The alternative seems to be that otherwise any provider MUST link against `libcrypto` to use these functions for handling `OSSL_PARAM` correctly in all its (underdocumented?) subtleties.

But forcing this seems to go against the original intent of not enforcing a `libcrypto` dependency for provider binaries.

**Comment by slontis** [comment]
> If it were me, I'd used the OSSL_PARAM_construct_xxx() unless a BIGNUM were involved. Then I'd use OSSL_PARAM_BLD_xxx(). I would never voluntarily choose the raw OSSL_PARAM approach.

I still dont agree with the way bignums are being stored.. If we get rid of this restriction and have a seperate type then you would not be forced into using the OSSL_PARAM_BLD either. I dont think being able to swap from a normal int and a bignum is worth the complexity.

**Comment by paulidale** [comment]
Even with a separate type, we could still do the swap.

**Comment by levitte** [comment]
> I still dont agree with the way bignums are being stored.. If we get rid of this restriction and have a seperate type then you would not be forced into using the OSSL_PARAM_BLD either. I dont think being able to swap from a normal int and a bignum is worth the complexity.

You would *still* have to deal with converting BIGNUMs to buffers and vice versa.  That is, unless you plan on passing the BIGNUM limbs directly (i.e. b->d), which would *force* everyone to use our BIGNUM implementation.

**Comment by mattcaswell** [comment]
> So, given this answer and the fact that any external provider author has to deal with OSSL_PARAM strings even for the most basic setup operations, shouldn't the various OSSL_PARAM_construct/BLD functions be included in core_dispatch?

No, I don't think they should be in core_dispatch. The core_dispatch functions should be about essential communication between providers and the core. These are just optional utility functions. You can use ours (link against libcrypto), or work with OSSL_PARAM directly, or create your own utilities. In most cases providers will just link against libcrypto. Our FIPS provider is relatively unusual in having a requirement to not link against libcrypto.

**Comment by levitte** [comment]
> In most cases providers will just link against libcrypto. Our FIPS provider is relatively unusual in having a requirement to not link against libcrypto.

Some HSMs have their own number implementation, I totally expect them to use that.  nCipher (now CHIL) is one of those, and [the engine](https://github.com/engine-corner/engine-chil/blob/master/e_chil.c) has to convert our BIGNUM to a buffer, and then feed their MPI API with it.

**Comment by romen** [comment]
@mattcaswell then we need to have carefully crafted documentation about the caveats of using the OSSL_PARAM directly, especially for utf8 strings and octet buffers.

At the moment I can't find enough info in our doc, examples use other APIs because OSSL_PARAM is cumbersome and our tests as well don't use the OSSL_PARAM API directly in most cases.


Re:
> The core_dispatch functions should be about essential communication between providers and the core. 

The most basic communication between providers and core happens through OSSL_PARAMs carrying utf8 strings, starting in the setup phase.

**Comment by levitte** [comment]
OSSL_PARAM.pod documents this type. Feel free to extend it.

**Comment by mattcaswell** [comment]
> Some HSMs have their own number implementation, I totally expect them to use that. nCipher (now CHIL) is one of those, and the engine has to convert our BIGNUM to a buffer, and then feed their MPI API with it.

Yes, of course. Although using a different number implementation doesn't prevent you from also linking against libcrypto (you'd have to hand the external number <-> OSSL_PARAM case yourself). In any case I do not mean to suggest that there aren't cases where providers don't want to link against libcrypto. It is clearly a requirement that it should be possible to write a provider without having to do that.

**Comment by paulidale** [comment]
> Some HSMs have their own number implementation, I totally expect them to use that.

Which is fine, they have to convert to their internal format regardless.

As we currently stand *everything* is forced to do a conversion.

**Comment by levitte** [comment]
> As we currently stand everything is forced to do a conversion.

Yup. Regardless

**Comment by romen** [comment]
Ok, the original issue is being addressed in #12389 so this can be closed.
